components/data-services/org.wso2.carbon.dataservices.capp.deployer/src/main/java/org/wso2/carbon/dataservices/capp/deployer/DataServiceCappDeployer.java

```diff
@@ -0,0 +1,162 @@
+
+/**
+ * This class is the implementation of the data service deployer.
+ */
+public class DataServiceCappDeployer implements AppDeploymentHandler {
+
+    private static final Log log = LogFactory.getLog(DataServiceCappDeployer.class);
+    private static final String DS_TYPE = "service/dataservice";
+    private static final String DS_DIR = "dataservices";
+
+    /**
+     * Deploy the data service artifacts and add them to data services.
+     *
+     * @param carbonApp  find artifacts from this CarbonApplication instance.
+     * @param axisConfig AxisConfiguration of the current tenant.
+     */
+    @Override
+    public void deployArtifacts(CarbonApplication carbonApp, AxisConfiguration axisConfig) throws DeploymentException {
+        if (log.isDebugEnabled()) {
+            log.debug("Deploying data services of carbon application - " + carbonApp.getAppName());
+        }
+        ApplicationConfiguration appConfig = carbonApp.getAppConfig();
+        List<Artifact.Dependency> dependencies = appConfig.getApplicationArtifact().getDependencies();
+
+        List<Artifact> artifacts = new ArrayList<>();
+        for (Artifact.Dependency dependency : dependencies) {
+            if (dependency.getArtifact() != null) {
+                artifacts.add(dependency.getArtifact());
+            }
+        }
+        deployUnDeployDataSources(true, artifacts, axisConfig);
+    }
+
+    /**
+     * Un-deploy the data services and remove them from data services.
+     *
+     * @param carbonApp  find artifacts from this CarbonApplication instance.
+     * @param axisConfig AxisConfiguration of the current tenant.
+     */
+    @Override
+    public void undeployArtifacts(CarbonApplication carbonApp, AxisConfiguration axisConfig)
+            throws DeploymentException {
+        if (log.isDebugEnabled()) {
+            log.debug("Un-Deploying data services of carbon application - " + carbonApp.getAppName());
+        }
+        ApplicationConfiguration appConfig = carbonApp.getAppConfig();
+        List<Artifact.Dependency> dependencies = appConfig.getApplicationArtifact().getDependencies();
+
+        List<Artifact> artifacts = new ArrayList<>();
+        for (Artifact.Dependency dependency : dependencies) {
+            if (dependency.getArtifact() != null) {
+                artifacts.add(dependency.getArtifact());
+            }
+        }
+        deployUnDeployDataSources(false, artifacts, axisConfig);
+    }
+
+    /**
+     * Deploy or un-deploy data services.
+     * if deploying, adding the data service to the data services.
+     * if un-deploying, removing the data service from data services.
+     * there can be multiple data sources as separate xml files.
+     *
+     * @param deploy    identify whether deployment process or un-deployment process.
+     * @param artifacts list of artifacts to be deployed.
+     */
+    private void deployUnDeployDataSources(boolean deploy, List<Artifact> artifacts, AxisConfiguration axisConfig)
```

Comment:
I suppose we should not have a common method for both deploying and undeploying datasources. We know which method is being called at the compilation time itself. Hence shall we extract the common code to aseparate method and break the 2 logics into 2 methods?

Reply:
Before creating this class discussed this and decided to create this class using the same structure of DataSourceCappDeployer class.

Reply:
Fixed in https://github.com/wso2/carbon-data/pull/254/files